src/native/eventpipe/ep-string.c

```diff
@@ -0,0 +1,106 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "ep-rt-config.h"
+
+#ifdef ENABLE_PERFTRACING
+#if !defined(EP_INCLUDE_SOURCE_FILES) || defined(EP_FORCE_INCLUDE_SOURCE_FILES)
+
+#include "ep-rt.h"
+#include <minipal/utf8.h>
+
+ep_char16_t *
+ep_rt_utf8_to_utf16le_string (
+	const ep_char8_t *str,
+	size_t len)
+{
+	if (!str)
+		return NULL;
+
+	if (len == (size_t) -1)
+		len = strlen(str);
+
+	if (len == 0) {
+		// Return an empty string if the length is 0
+		ep_char16_t * empty_str = ep_rt_utf16_string_alloc(1);
+		if(empty_str == NULL)
+			return NULL;
+
+		*empty_str = '\0';
+		return empty_str;
+	}
+
+	int32_t flags = MINIPAL_MB_NO_REPLACE_INVALID_CHARS | MINIPAL_TREAT_AS_LITTLE_ENDIAN;
+	size_t ret = minipal_get_length_utf8_to_utf16 (str, len, flags);
+	if (ret <= 0)
+		return NULL;
+
+	ep_char16_t * converted_str = ep_rt_utf16_string_alloc(ret + 1);
+	if (converted_str == NULL)
+		return NULL;
+
+	ret = minipal_convert_utf8_to_utf16 (str, len, (CHAR16_T *)converted_str, ret, flags);
+	converted_str[ret] = '\0';
+	return converted_str;
+}
+
+static
+ep_char8_t *
+ep_utf16_to_utf8_string_impl (
+	const ep_char16_t *str,
+	size_t len,
+	bool treat_as_le)
+{
+	if (!str)
+		return NULL;
+
+	if (len == (size_t) -1)
```

Comment:
This contract where `(size_t) -1` means "compute the length for me" is not great. Can we replace it by a new method that does not take the length argument and computes the C string length for you, and replace all callsites that pass -1 to call this method instead?

Reply:
If we are doing this, I think we can also try to rework `// Return an empty string if the length is 0` at the same time; null check at call sites instead of special casing it in the helper. If it is possible, then we can remove these wrappers altogether and inline minipal_* at the call sites.

Reply:
I looked at the empty string handling a bit. Trying to move it out seemed troublesome enough that I'd rather leave the empty string conversion in the helper.